### PR TITLE
Adding parameter check for 

### PR DESCRIPTION
**Pull request by allenjzhang**
TrackedResource
GenericeResource
GenericeResourceData
GenericeResourceOperation
OperationBase
ResourceOperationBase

### CODE REVIEW

File: Azure.ResourceManager.Core/OperationsBase.cs [hunk 1]
```diff
@@ -67,7 +75,7 @@ protected OperationsBase(AzureResourceManagerClientOptions options, ResourceIden
         protected virtual void Validate(ResourceIdentifier identifier)
         {
             if (identifier?.Type != ValidResourceType)
-                throw new InvalidOperationException($"Invalid resource type {identifier?.Type} expected {ValidResourceType}");
+                throw new ArgumentException($"Invalid resource type {identifier?.Type} expected {ValidResourceType}");
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/OperationsBase.cs:78]: nameof parameter

File: Azure.ResourceManager.Core/GenericResourceData.cs [hunk 2]
```diff
@@ -16,11 +16,16 @@ public class GenericResourceData : TrackedResource<ResourceManager.Resources.Mod
...
         public GenericResourceData(ResourceManager.Resources.Models.GenericResource genericResource)
-            : base(genericResource.Id, genericResource.Location, genericResource)
+            : base(genericResource?.Id, genericResource?.Location, genericResource)
         {
+            if (genericResource is null)
+                throw new ArgumentNullException(nameof(genericResource));
+
             Tags = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase);
-            Tags.Clear();
             foreach (var tag in genericResource.Tags)
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/GenericResourceData.cs:29]: Add extension.

File: Azure.ResourceManager.Core/GenericResourceData.cs [hunk 3]
```diff
@@ -89,6 +94,10 @@ public GenericResourceData(ResourceIdentifier id, LocationData location)
         /// <param name="other"> The tracked resource convert from. </param>
         public static implicit operator ResourceManager.Resources.Models.GenericResource(GenericResourceData other)
         {
+            if (other == null)
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/GenericResourceData.cs:97]: is

File: Azure.ResourceManager.Core/GenericResourceData.cs [hunk 4]
```diff
@@ -89,6 +94,10 @@ public GenericResourceData(ResourceIdentifier id, LocationData location)
...
+                return null;
+
+            // ?
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/GenericResourceData.cs:100]: Add comment: Temp code. Following block should not be codegen.

File: Azure.ResourceManager.Core/GenericResourceOperations.cs [hunk 5]
```diff
@@ -122,33 +129,25 @@ public async Task<ArmResponse<GenericResource>> AddTagAsync(string key, string v
...
-        /// Add a tag to the resource.
-        /// </summary>
-        /// <param name="key"> The tag key. </param>
-        /// <param name="value"> The tag value. </param>
-        /// <returns>An <see cref="ArmOperation{TOperations}"/> that allows the user to control polling and waiting for Tag completion.</returns>
+        /// <inheritdoc/>
         public ArmOperation<GenericResource> StartAddTag(string key, string value)
         {
+            if (string.IsNullOrWhiteSpace(key))
+                throw new ArgumentException(nameof(key));
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/GenericResourceOperations.cs:136]: add messages back to all ArgumentExeption

File: Azure.ResourceManager.Core/GenericResourceOperations.cs [hunk 6]
```diff
@@ -176,11 +175,19 @@ public override async Task<ArmResponse<GenericResource>> GetAsync(CancellationTo
         /// <inheritdoc/>
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/GenericResourceOperations.cs:175]: > /// <inheritdoc/> [](start = 7, length = 18)

update test case and uncomment out the check

File: Azure.ResourceManager.Core/Adapters/TrackedResource.cs [hunk 7]
```diff
@@ -20,6 +21,13 @@ public abstract class TrackedResource<TModel> : TrackedResource
         /// <param name="data"> The model to copy from. </param>
         protected TrackedResource(ResourceIdentifier id, LocationData location, TModel data)
         {
+            if (id is null)
+                throw new ArgumentNullException(nameof(id));
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/Adapters/TrackedResource.cs:25]: id could be null. Create unit tests

File: Azure.ResourceManager.Core/GenericResourceOperations.cs [hunk 8]
```diff
@@ -114,6 +118,9 @@ public ArmResponse<GenericResource> AddTag(string key, string value)
         /// Tag completion. </returns>
         public async Task<ArmResponse<GenericResource>> AddTagAsync(string key, string value, CancellationToken cancellationToken = default)
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/GenericResourceOperations.cs:119]: Double check value null causing issue and give default of string.Empty.

File: Azure.ResourceManager.Core/Utils/UtilityExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Azure.ResourceManager.Core.Utils
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/Utils/UtilityExtensions.cs:5]: namespace Azure.ResourceManager.Core

File: Azure.ResourceManager.Core/GenericResourceOperations.cs [hunk 10]
```diff
@@ -234,7 +212,7 @@ public ArmResponse<GenericResource> RemoveTag(string key)
         public async Task<ArmResponse<GenericResource>> RemoveTagAsync(string key, CancellationToken cancellationToken = default)
         {
             GenericResource resource = GetResource();
-            DeleteTag(key, resource.Data.Tags);
+            resource.Data.Tags.Remove(key);
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/GenericResourceOperations.cs:215]: if false, no key found, skip the call and return resource.

File: Azure.ResourceManager.Core/GenericResourceOperations.cs [hunk 11]
```diff
@@ -87,70 +88,42 @@ public async Task<ArmOperation<Response>> StartDeleteAsync(CancellationToken can
...
-        /// </summary>
-        /// <param name="key"> The tag key. </param>
-        /// <param name="value"> The tag value. </param>
-        /// <returns>An <see cref="ArmResponse{TOperations}"/> that allows the user to control polling and waiting for Tag completion.</returns>
+        /// <inheritdoc/>
         public ArmResponse<GenericResource> AddTag(string key, string value)
         {
             GenericResource resource = GetResource();
-            UpdateTags(key, value, resource.Data.Tags);
+            resource.Data.Tags[key] = value;
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/GenericResourceOperations.cs:95]: Add a comment optimization, if same key exist with same value, skip call.

File: Azure.ResourceManager.Core/OperationsBase.cs [hunk 12]
```diff
@@ -19,8 +19,16 @@ public abstract class OperationsBase
         /// <param name="id"> The identifier of the resource that is the target of operations. </param>
         /// <param name="credential"> A credential used to authenticate to an Azure Service. </param>
         /// <param name="baseUri"> The base URI of the service. </param>
+        /// <exception cref="ArgumentNullException">
+        ///     <paramref name="options"/> or <paramref name="credential"/> is null.
+        /// </exception>
         protected OperationsBase(AzureResourceManagerClientOptions options, ResourceIdentifier id, TokenCredential credential, Uri baseUri)
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/OperationsBase.cs:25]: >credential [](start = 115, length = 10)

rollback

File: Azure.ResourceManager.Core/ResourceOperationsBase.cs [hunk 13]
```diff
@@ -19,7 +19,7 @@ public abstract class ResourceOperationsBase : OperationsBase
         /// </summary>
         /// <param name="operations"> The operations representing the resource. </param>
         protected ResourceOperationsBase(ResourceOperationsBase operations)
-            : base(operations.ClientOptions, operations.Id, operations.Credential, operations.BaseUri)
+            : base(operations?.ClientOptions, operations?.Id, operations?.Credential, operations?.BaseUri)
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/ResourceOperationsBase.cs:22]: >ClientOptions [](start = 31, length = 13)

revisit intention. Rollback.

**Review by m-nash — APPROVED** [review]

**Pull request merged**